Add documentation for nested search in knn (#6237)
opensearch-project/documentation-website: _search-plugins/knn/nested-search-knn.md

---
layout: default
title: k-NN search with nested fields
nav_order: 21
parent: k-NN search
grand_parent: Search methods
has_children: false
has_math: true
---

# k-NN search with nested fields
With k-NN search using nested fields, you can store multiple vectors in a single document and search for the k-nearest documents. 

> Each document is scored using the nearest vector in the nested field. 

Reviewer: this phase is a bit misleading. It doesn't say if the whole document is scored based on a best score from all the vectors from its nested field(s). Need more details on what is `nearest vector` in the context of this sentence.

Author: Right. It could be or could not be the best score from all the vectors from its nested field. Let me rephrase it.

---
layout: default
title: k-NN search with nested fields
nav_order: 21
parent: k-NN search
grand_parent: Search methods
has_children: false
has_math: true
---

# k-NN search with nested fields
With k-NN search using nested fields, you have the flexibility to store multiple vectors within a single document.
For instance, if your document comprises various components, you can generate a vector value for each component and store them in a nested field within the document.

> When performing a search with a k value of 2, even if a single document has 2 nearest vectors, the search results will include 2 documents: one with the nearest vector and another with the third nearest vector.

Reviewer: Can this sentence be rephrased into a more simpler form? Maybe first state the overall high-level idea that the document will be included to the result only once and that is based on the highest score/most similarity of all vectors stored in all its nested fields. Example will follow, something like: index has 2 docs, doc A has two fields with vector A1 and A2, and doc B with vector B1. When we run knn search with k == 2. Most similar vectors are A1, A2, B1, but final result will be [A, B] because each doc id will be unique.

---
layout: default
title: k-NN search with nested fields
nav_order: 21
parent: k-NN search
grand_parent: Search methods
has_children: false
has_math: true
---

# k-NN search with nested fields

> Using [nested fields]({{site.url}}{{site.baseurl}}/field-types/nested/) in a k-NN index, you can store multiple vectors in a single document. For example, if your document consists of various components, you can generate a vector value for each component and store each vector in a nested field.

Reviewer 2: Should we define k-NN on first appearance in this para?

> Using [nested fields]({{site.url}}{{site.baseurl}}/field-types/nested/) in a k-NN index, you can store multiple vectors in a single document. For example, if your document consists of various components, you can generate a vector value for each component and store each vector in a nested field.

Reviewer 3: ```suggestion
Using [nested fields]({{site.url}}{{site.baseurl}}/field-types/nested/) in a k-nearest neighbors (k-NN) index, you can store multiple vectors in a single document. For example, if your document consists of various components, you can generate a vector value for each component and store each vector in a nested field.
```

> A k-nearest neighbors document search operates at the field level. For a document with nested fields, OpenSearch examines only the vector nearest to the query vector to decide whether to include the document in the results. For example, consider an index containing documents `A` and `B`. Document `A` is represented by vectors `A1` and `A2`, and document `B` is represented by vector `B1`. Further, the similarity order for a query Q is `A1`, `A2`, `B1`. If you search using query Q with a k value of 2, the search will return both documents `A` and `B` instead of only document `A`.

Reviewer 2: ```suggestion
A k-NN document search operates at the field level. For a document with nested fields, OpenSearch examines only the vector nearest to the query vector to decide whether to include the document in the results. For example, consider an index containing documents `A` and `B`. Document `A` is represented by vectors `A1` and `A2`, and document `B` is represented by vector `B1`. Further, the similarity order for a query Q is `A1`, `A2`, `B1`. If you search using query Q with a k value of 2, the search will return both documents `A` and `B` instead of only document `A`.
```

Note that in the case of an approximate search, the results are approximations and not exact matches.

> k-NN search with nested fields is supported by the HNSW algorithm for the Lucene and the Faiss engines. 

Reviewer 2: ```suggestion
k-NN search with nested fields is supported by the HNSW algorithm for the Lucene and Faiss engines. 
```

```json
PUT my-knn-index-1
{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "nested_field": {
        "type": "nested",
        "properties": {
          "my_vector": {
            "type": "knn_vector",
            "dimension": 3,
            "method": {
              "name": "hnsw",
              "space_type": "l2",
              "engine": "lucene",
              "parameters": {
                "ef_construction": 100,
                "m": 16
              }
            }
          }
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

After you create the index, add some data to it:

```json
PUT _bulk?refresh=true
{ "index": { "_index": "my-knn-index-1", "_id": "1" } }
{"nested_field":[{"my_vector":[1,1,1]},{"my_vector":[2,2,2]},{"my_vector":[3,3,3]}]}
{ "index": { "_index": "my-knn-index-1", "_id": "2" } }
{"nested_field":[{"my_vector":[10,10,10]},{"my_vector":[20,20,20]},{"my_vector":[30,30,30]}]}
```
{% include copy-curl.html %}

> Then run a k-nearest neighbor search on the data using the `knn` query type:

Reviewer 2: ```suggestion
Then run a k-NN search on the data by using the `knn` query type:
```

```json
GET my-knn-index-1/_search
{
  "query": {
    "nested": {
      "path": "nested_field",
      "query": {
        "knn": {
          "nested_field.my_vector": {
            "vector": [1,1,1],
            "k": 2
          }
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

> Even though all three nearest vectors to the query vector are in document 1, the query returns both documents 1 and 2 because k is set to 2:

Reviewer 2: ```suggestion
Even though all three vectors nearest to the query vector are in document 1, the query returns both documents 1 and 2 because k is set to 2:
```

      {
        "_index": "my-knn-index-1",
        "_id": "1",
        "_score": 1,
        "_source": {
          "nested_field": [
            {
              "my_vector": [
                1,
                1,
                1
              ]
            },
            {
              "my_vector": [
                2,
                2,
                2
              ]
            },
            {
              "my_vector": [
                3,
                3,
                3
              ]
            }
          ]
        }
      },
      {
        "_index": "my-knn-index-1",
        "_id": "2",
        "_score": 0.0040983604,
        "_source": {
          "nested_field": [
            {
              "my_vector": [
                10,
                10,
                10
              ]
            },
            {
              "my_vector": [
                20,
                20,
                20
              ]
            },
            {
              "my_vector": [
                30,
                30,
                30
              ]
            }
          ]
        }
      }
    ]
  }
}
```

## k-NN search with filtering on nested fields

> You can apply a filter to k-NN search with nested fields. A filter can be applied to either a top-level field or a field inside a nested field.

Reviewer 2: ```suggestion
You can apply a filter to a k-NN search with nested fields. A filter can be applied to either a top-level field or a field inside a nested field.
```

{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "nested_field": {
        "type": "nested",
        "properties": {
          "my_vector": {
            "type": "knn_vector",
            "dimension": 3,
            "method": {
              "name": "hnsw",
              "space_type": "l2",
              "engine": "lucene",
              "parameters": {
                "ef_construction": 100,
                "m": 16
              }
            }
          }
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

After you create the index, add some data to it:

```json
PUT _bulk?refresh=true
{ "index": { "_index": "my-knn-index-1", "_id": "1" } }
{"parking": false, "nested_field":[{"my_vector":[1,1,1]},{"my_vector":[2,2,2]},{"my_vector":[3,3,3]}]}
{ "index": { "_index": "my-knn-index-1", "_id": "2" } }
{"parking": true, "nested_field":[{"my_vector":[10,10,10]},{"my_vector":[20,20,20]},{"my_vector":[30,30,30]}]}
{ "index": { "_index": "my-knn-index-1", "_id": "3" } }
{"parking": true, "nested_field":[{"my_vector":[100,100,100]},{"my_vector":[200,200,200]},{"my_vector":[300,300,300]}]}
```
{% include copy-curl.html %}

> Then run a k-nearest neighbors search on the data using the `knn` query type with a filter. The following query returns documents whose `parking` field is set to `true`:

Reviewer 2: ```suggestion
Then run a k-NN search on the data using the `knn` query type with a filter. The following query returns documents whose `parking` field is set to `true`:
```

```json
GET my-knn-index-1/_search
{
  "query": {
    "nested": {
      "path": "nested_field",
      "query": {
        "knn": {
          "nested_field.my_vector": {
            "vector": [
              1,
              1,
              1
            ],
            "k": 3,
            "filter": {
              "term": {
                "parking": true
              }
            }
          }
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

> Even though all three nearest vectors to the query vector are in document 1, the query returns documents 2 and 3 because document 1 is filtered out:

Reviewer 2: ```suggestion
Even though all three vectors nearest to the query vector are in document 1, the query returns documents 2 and 3 because document 1 is filtered out:
```
